File: C++/017. Letter Combinations of a Phone Number.cpp
class Solution {
public:
    vector<string> letters;
    Solution()
    {
       letters = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    }
    
    vector<string> letterCombinations(string digits) 
    {
        int n = digits.length();
        if (n == 0) return {};
        int num =  digits.back() - '0';
        digits.pop_back();
        vector<string> prev = letterCombinations(digits);
        vector<string> res = {};
        if (prev.size() == 0)
        {
            prev = {""};
        }
        for(int i = 0; i < prev.size(); i++)
        {
            for (int j = 0; j < letters[num-2].size(); j++)
            {
                res.push_back(prev[i] + letters[num-2][j]);
            }
        }
            
        return res;
    }
};

File: python/017. Letter Combinations of a Phone Number.py
class Solution(object):
    def dfs(self, digits, idx, cur, res):
        if idx == len(digits):
            res.append(copy.copy(cur))
            return
        for c in self.status[int(digits[idx])]:
            cur += c
            self.dfs(digits, idx+1, cur, res)
            cur = cur[:len(cur)-1]
        return
    
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if len(digits) == 0: return []
        self.status = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
        cur = ""
        res = []
        self.dfs(digits, 0, cur, res)
        return res